src/miral/simulated_secondary_click.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "miral/simulated_secondary_click.h"
+
+#include "mir/options/option.h"
+#include "mir/server.h"
+#include "mir/shell/accessibility_manager.h"
+#include "mir/shell/simulated_secondary_click_transformer.h"
+#include "mir/synchronised.h"
+
+struct miral::SimulatedSecondaryClick::Self
+{
+    explicit Self(bool enabled_by_default) :
+        state{State{enabled_by_default}}
+    {
+    }
+
+    std::weak_ptr<mir::shell::AccessibilityManager> accessibility_manager;
+
+    struct State
+    {
+        explicit State(bool enabled) :
+            enabled{enabled}
+        {
+        }
+
+        bool enabled;
+        std::chrono::milliseconds hold_duration{1000};
+        float displacement_threshold{20.0f};
+        std::function<void()> on_hold_start, on_hold_cancel, on_secondary_click;
```

Comment:
We tend not to define multiple variables in a statement.
Initializing these with an empty implementation would avoid the need to test for "callable".

```suggestion
        std::function<void()> on_hold_start{[]{}};
        std::function<void()> on_hold_cancel{[]{}};
        std::function<void()> on_secondary_click{[]{}};
```